packages/server/graphql/mutations/endRetrospective.ts

```diff
@@ -0,0 +1,192 @@
+          .count()
+          .default(0) as unknown) as number,
+        topicCount: reflectionGroupIds.length,
+        reflectionCount: reflections.length
+      },
+      {nonAtomic: true}
+    )
+    .run()
+}
+
+export default {
+  type: new GraphQLNonNull(EndRetrospectivePayload),
+  description: 'Finish a retrospective meeting',
+  args: {
+    meetingId: {
+      type: new GraphQLNonNull(GraphQLID),
+      description: 'The meeting to end'
+    }
+  },
+  async resolve(_source, {meetingId}, context: GQLContext) {
+    const {authToken, socketId: mutatorId, dataLoader} = context
+    const r = await getRethink()
+    const operationId = dataLoader.share()
+    const subOptions = {mutatorId, operationId}
+    const now = new Date()
+    const viewerId = getUserId(authToken)
+
+    // AUTH
+    const meeting = (await r
+      .table('NewMeeting')
+      .get(meetingId)
+      .default(null)
+      .run()) as MeetingRetrospective | null
+    if (!meeting) return standardError(new Error('Meeting not found'), {userId: viewerId})
+    const {endedAt, facilitatorStageId, meetingNumber, phases, teamId, meetingType} = meeting
+
+    // VALIDATION
+    if (!isTeamMember(authToken, teamId) && authToken.rol !== 'su') {
+      return standardError(new Error('Team not found'), {userId: viewerId})
+    }
+    if (endedAt) return standardError(new Error('Meeting already ended'), {userId: viewerId})
+
+    // RESOLUTION
+    const currentStageRes = findStageById(phases, facilitatorStageId)
+    if (!currentStageRes) {
+      return standardError(new Error('Cannot find facilitator stage'), {userId: viewerId})
+    }
+    const {stage} = currentStageRes
+    const phase = getMeetingPhase(phases)
+    stage.isComplete = true
+    stage.endAt = now
+
+    const completedRetrospective = ((await r
+      .table('NewMeeting')
+      .get(meetingId)
+      .update(
+        {
+          endedAt: now,
+          phases
+        },
+        {returnChanges: true}
+      )('changes')(0)('new_val')
+      .run()) as unknown) as MeetingRetrospective
+
+    // remove any empty tasks
+    const removedTaskIds = await removeEmptyTasks(meetingId)
+
+    const [meetingMembers, team] = await Promise.all([
+      dataLoader.get('meetingMembersByMeetingId').load(meetingId),
+      dataLoader.get('teams').load(teamId)
+    ])
+    const presentMembers = meetingMembers.filter(
+      (meetingMember) => meetingMember.isCheckedIn === true
+    )
+    const presentMemberUserIds = presentMembers.map(({userId}) => userId)
+    endSlackMeeting(meetingId, teamId, dataLoader).catch(console.log)
+    finishRetroMeeting(completedRetrospective, dataLoader)
+    const {facilitatorUserId, templateId} = completedRetrospective
+    const template = await dataLoader.get('meetingTemplates').load(templateId)
+    const {name: meetingTemplateName} = template
+    presentMemberUserIds.forEach((userId) => {
+      const wasFacilitator = userId === facilitatorUserId
+      segmentIo.track({
```

Comment:
duplicate code 3 times & it's probably time to refactor...

can we refactor this so the majority of the logic is shared across meetings? maybe we pass in a meeting object to a function called `sendMeetingEndToSegment` or similar